### PR TITLE
Update parabolic model based TOF to d-spacing calculation

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

- Related Gitlab Defect: [PD269](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/269)
- This is #31736 into `master`

This PR addresses an issue related to converting from TOF to d-spacing when using a parabolic model.
More specifically, non-realistic d-spacing can be picked up using previous selection criteria.
The new criteria will always select the positive root that is closer to 0.0.

**To test:**

> The testing instruction is derived from the related defect
- Go to analysis cluster
- Git clone [mantid-total-scattering](https://github.com/neutrons/mantid_total_scattering)
- Checkout this branch and build mantid workbench
- Inside workbench, use the following script
```python
#!/usr/bin/env python

import sys
sys.path.append("~/mantid_total_scattering")

import json
from total_scattering.reduction import TotalScatteringReduction

# calibration file d143068_2020_02_11
jsonfile = "/SNS/NOM/shared/User_story_test/NOM_US-269/mantidts_test/ceriaBulk_PAC06_at_10K_IDL_Calib.json"
with open(jsonfile, 'r') as jf:
    config = json.load(jf)

# execute
TotalScatteringReduction(config)
```
Instead of triggering a runtime error, the reduction should finished within about 1h or so.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by wdzhou — COMMENTED** [review]
Except one question about negative spacing value, everything looks good to me.

File: Framework/Kernel/src/Unit.cpp
Comment (KedoKudo) [reply Framework/Kernel/src/Unit.cpp:712]: @wdzhou 
The edge case for two negative roots are handled above with the following checking
```c++
  if ((difa > 0) && ((tzero - tof) > 0)) {
    throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't "
                             "have a positive root");
  }
```

Comment (wdzhou) [reply Framework/Kernel/src/Unit.cpp:712]: Suggestion:  From https://github.com/mantidproject/mantid/blob/PD269_binningErrorInAlignAndFocusPowder/Framework/Kernel/src/Unit.cpp#L661 to https://github.com/mantidproject/mantid/blob/PD269_binningErrorInAlignAndFocusPowder/Framework/Kernel/src/Unit.cpp#L671, how about if () {} else if () {} else if () ... ?

Comment (wdzhou) [reply Framework/Kernel/src/Unit.cpp:712]: And also, it is better to add some doc for `DIFA < 0` such that in this case, there must at least 1 positive root?

Comment (KedoKudo) [reply Framework/Kernel/src/Unit.cpp:712]: - For the edge cases, the current ones are functionally equivalent to the suggested if-elif structure, but cleaner. Therefore I don't believe it is necessary to make the change.
- For the second question regarding add docs on the sign of `DIFA`, I would advise again doing so as it might lead the future developers down the path that was causing us the trouble in the first place.  More specially, the associated issue was due to the fact that someone was adapting the solution based on the documentation of deprecated [AlignDetectors](https://docs.mantidproject.org/nightly/algorithms/AlignDetectors-v1.html), which used sign of `DIFA` to determine the root.

**Review by wdzhou — APPROVED** [review]
It looks fine to me.

**Comment by DavidFair** [comment]
@KedoKudo Is this the same as #31734 ? 
If so, is it worth closing one of these to not duplicate reviews / effort?  This saves you having to make identical changes, or cherry-pick commits into both.

Also I'm happy to potentially merge this into `master` but I would have thought you prefer this going into `ornl-next` (which gets merged back into master periodically)? It's completely your call

**Comment by KedoKudo** [comment]
> @KedoKudo Is this the same as #31734 ?
> If so, is it worth closing one of these to not duplicate reviews / effort? This saves you having to make identical changes, or cherry-pick commits into both.
> 
> Also I'm happy to potentially merge this into `master` but I would have thought you prefer this going into `ornl-next` (which gets merged back into master periodically)? It's completely your call

@DavidFair It could be great if you can merge this into `master`.
The other one, #31734 , is the same branch with a different merging target, and I can merge it myself once this one is merged into `master`. :smile:

**Review by DavidFair — APPROVED** [review]

**Comment by DavidFair** [comment]
> The other one, #31736 , is the same branch with a different merging target, and I can merge it myself once this one is merged into `master`. smile

Done, I can never quite remember the ORNL workflow so worth a check :laughing:

**Pull request merged**
